tests_e2e/scenarios/lib/logging.py

```diff
@@ -14,24 +14,110 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-import logging
 
 #
-# This module defines a single object, 'log', which test use for logging.
-#
-# When the test is invoked as part of a LISA test suite, 'log' references the LISA root logger.
-# Otherwise, it references a new Logger named 'waagent'.
+# This module defines a single object, 'log', of type AgentLogger, which the end-to-end tests and libraries use
+# for logging.
 #
 
-log: logging.Logger = logging.getLogger("lisa")
+from logging import FileHandler, Formatter, Handler, Logger, StreamHandler, INFO
+from pathlib import  Path
+from threading import current_thread
+from typing import Dict, Callable
+
+
+class _AgentLoggingHandler(Handler):
+    """
+    AgentLoggingHandler is a helper class for AgentLogger.
+
+    This handler simply redirects logging to other handlers. It maintains a set of FileHandlers associated to specific
+    threads. When a thread emits a log record, the AgentLoggingHandler passes through the call to the FileHandlers
+    associated with that thread, or to a StreamHandler that outputs to stdout if there is not a FileHandler for that
+    thread.
+
+    Thread can set a FileHandler for themselves using _AgentLoggingHandler.set_current_thread_log() and remove that
+    handler using _AgentLoggingHandler.close_current_thread_log().
+
+    The _AgentLoggingHandler simply passes through calls to setLevel, setFormatter, flush, and close to the handlers
+    it maintains.
+
+    AgentLoggingHandler is meant to be primarily used in multithreaded scenarios and is thread-safe.
+    """
+    def __init__(self):
+        super().__init__()
+        self.formatter: Formatter = Formatter('%(asctime)s.%(msecs)03d [%(levelname)s] %(message)s', datefmt="%Y-%m-%dT%H:%M:%SZ")
+        self.default_handler = StreamHandler()
+        self.default_handler.setFormatter(self.formatter)
+        self.per_thread_handlers: Dict[int, FileHandler] = {}
+
+    def set_thread_log(self, thread_ident: int, log_file: Path) -> None:
```

Comment:
How is the thread_ident determined?

Reply:
It is a property of Python's Thread objects. 

This method is meant to be used by tests that start multiple threads (currently we don't many of those) and want the child threads to write to the test's log file. The parent thread would create a new Thread, set the log file using Thread.ident and then start the child thread with Thread.start(). The child thread should call close_current_thread_log() when it completes. I'll document this protocol in the wiki.

Note that _AgentLoggingHandler is not a public class and is not meant to be used outside this module; it is simply an implementation detail of the AgentLogger. Users of the logger would call set_thread_log(), etc on the logger, which simply forwards to the _AgentLoggingHandler.